### ISSUE TITLE
Nested Custom Coercers Not Applied

### ISSUE THREAD

**Issue by sdorazio** [pyeve/cerberus#325]
Used Cerberus version / latest commit: Cerberus==1.1

- I consulted these documentations:
  - [x] http://docs.python-cerberus.org/en/stable/
  - [x] http://docs.python-cerberus.org/en/latest/

- I consulted these sections of the docs (add more lines as necessary):
  - ...
  - ...

- [ ] I found nothing relevant to my problem in the docs.
- [ ] I found the documentation not helpful to my problem.

- [ ] I have the capacity to improve the docs when my problem is solved.
- [ ] I have the capacity to submit a patch when a bug is identified.

----

#### Support request / Bug report

I am trying to apply a custom coercer to a nested item. I do not believe this is working properly. To simplify this, I modified the example code from the [documentation](http://docs.python-cerberus.org/en/stable/customize.html#custom-coercers) to demonstrate the issue:

```
class MyNormalizer(Validator):
    def __init__(self, multiplier, *args, **kwargs):
        super(MyNormalizer, self).__init__(multiplier=multiplier, *args, **kwargs)
        self.multiplier = multiplier

    def _normalize_coerce_multiply(self, value):
        return value * self.multiplier

schema = {
    'foo': {
        'type': 'list',
        'schema': {
            'sub_foo': {
                'coerce': 'multiply'
            }
        }
    }
}

document = {'foo': [{'sub_foo': 2}]}
MyNormalizer(2).normalized(document, schema)

# Expected Output: {'foo': [{'sub_foo': 4}]}
# Output: {'foo': [{'sub_foo': 2}]}
```

I believe the schema definition / custom `Validator` class is accurate. Please let me know if this is not the case.

**Comment by fermuch** [comment]
I just had the same problem with a nested `dict`.
Seems like `coerce` methods aren't called if they're inside a `dict` or `list`.

My schema, for reference:

```python
import arrow
from cerberus import Validator

date_metatag = {
    'type': 'dict',
    'dependencies': {'type': 'date'},
    'schema': {
        'date': {
            'type': 'datetime',
            'required': True,
            'coerce': lambda x: arrow.get(x).datetime
        }
    }
}

metatag_schema = {
    'type': {
        'type': 'string',
        'required': True,
        'allowed': [
            'date'
        ]
    },
    'value': {
        'required': True,
        'type': 'dict',
        'oneof': [
            date_metatag
        ]
    }
}

v = Validator(metatag_schema)
v.validated({
    'type': 'date',
    'value': {
        'date': '2017-07-14T20:05:53.805Z'
    }
})import arrow
from cerberus import Validator

date_metatag = {
    'type': 'dict',
    'dependencies': {'type': 'date'},
    'schema': {
        'date': {
            'type': 'datetime',
            'required': True,
            'coerce': lambda x: arrow.get(x).datetime
        }
    }
}

metatag_schema = {
    'type': {
        'type': 'string',
        'required': True,
        'allowed': [
            'date'
        ]
    },
    'value': {
        'required': True,
        'type': 'dict',
        'oneof': [
            date_metatag
        ]
    }
}

v = Validator(metatag_schema)
v.validated({
    'type': 'date',
    'value': {
        'date': '2017-07-14T20:05:53.805Z'
    }
})

v.errors:
>>> {'value': [{'oneof': ['none or more than one rule validate', {'oneof definition 0': ['must be of datetime type']}]}]}
```

**Comment by funkyfuture** [comment]
@sdorazio if i'm not mistaken your `schema` rule's constraint aren't appropriate for a usage on  a [list](http://docs.python-cerberus.org/en/stable/validation-rules.html#schema-list).

**Comment by sireliah** [comment]
Hey, I had the same problem with `coerce` and `rename` inside nested structures.

```
from cerberus import Validator

schema = {
    "rows": {
        "type": "list",

        "schema": {
            "type": "dict",
            "schema": {
                "reference": {
                    "type": "string",
                    "rename": "REFERENCE"
                },
            }
        }
    }
}


to_normalize = {"rows": {"reference": "aaaaa"}}

v = Validator(schema)

v.normalized(to_normalize)

Out[18]: {'rows': {'reference': 'aaaaa'}}
```

In this example the nested data is not normalized. If there is no easy fix for that issue, I can probably dig in the code and try to find what fails.

**Comment by sireliah** [comment]
Hey, I found similar issue with nested structures when you:
- rename the parent element 
- try to change child element

```python
from cerberus import Validator

schema = {
    "rows": {
        "type": "list",
        "rename": "ROWS",
        "schema": {
            "type": "dict",
            "schema": {
                "reference": {
                    "type": "string",
                    "coerce": lambda x: "%s MODIFIED HERE " % x
                },
            }
        }
    }
}


to_normalize = {"rows": [{"reference": "aaaaa"}]}

v = Validator(schema)

v.normalized(to_normalize)

Out[29]: {'ROWS': [{'reference': 'aaaaa MODIFIED HERE '}]}

Expected: {'ROWS': [{'reference': 'aaaaa'}]}
```

**Comment by funkyfuture** [comment]
@fermuch normalization rules in `*of` rules are not applied. see #344 / 17e786b

@sireliah at the time values are coerced, the document will not have a field with any rules in the schema. see #345

**Comment by funkyfuture** [comment]
i think this issue can be closed as #344 got merged.
